runtime/src/blueoil_image.cpp

```diff
@@ -100,26 +100,39 @@ Tensor ResizeHorizontal_BiLinear(const Tensor &tensor, const int width) {
   const int channels  = shape[2];
   const int height = srcHeight;
   Tensor dstTensor({height, width, channels});
-  float xScale = static_cast<float>(width) / static_cast<float>(srcWidth);
-  int xSrcWindow = std::floor(1/xScale);
-  xSrcWindow = (xSrcWindow < 2)? 2 :xSrcWindow;
+  const float xScale = static_cast<float>(width) / static_cast<float>(srcWidth);
+  const float xSrcWindow = (xScale < 1.0)?(1.0f / xScale):1.0;
   for (int dstY = 0 ; dstY < height ; dstY++) {
     for (int dstX = 0 ; dstX < width ; dstX++) {
-      int srcX = static_cast<int>(std::floor(dstX/xScale));
+      float srcXf = (dstX + 0.5)/xScale - 0.5;
       int srcY = dstY;
+      int xStart = std::ceil(srcXf - xSrcWindow);
+      int xEnd = std::floor(srcXf + xSrcWindow);
+      if (xStart >= xEnd) {  // for enlarge scale
+        xStart = std::floor(srcXf);
+        xEnd = std::ceil(srcXf);
+      }
+      // don't convolve pixels outside the frame
+      if (xStart < 0) {
+        xStart = 0;
+      }
+      if (xEnd >= srcWidth) {
+        xEnd = srcWidth - 1;
+      }
       for (int c = 0 ; c < channels ; c++) {
         float v = 0.0;
         float totalW = 0.0;
-        for (int x = -xSrcWindow ; x < xSrcWindow; x++) {
-          int srcX2 = clamp(srcX + x, 0, srcWidth - 1);
-          const float *srcRGB = tensor.dataAsArray({srcY, srcX2, 0});
-          float d = std::abs(static_cast<float>(x) / static_cast<float> (xSrcWindow));
-          float w = 1.0 - d;  // Bi-Linear
-          v += w * srcRGB[c];
-          totalW += w;
+        for (int srcX = xStart ; srcX <= xEnd; srcX++) {
+          float d = std::abs(static_cast<float>(srcX) - srcXf) / xSrcWindow;
+          if (d < 1.0) {
+            const float *srcPixel = tensor.dataAsArray({srcY, srcX, c});
+            float w = 1.0 - d;  // Bi-Linear
+            v += w * (*srcPixel);
+            totalW += w;
+          }
         }
-        float *dstRGB = dstTensor.dataAsArray({dstY, dstX, 0});
-        dstRGB[c] = v / totalW;
+        float *dstPixel = dstTensor.dataAsArray({dstY, dstX, c});
+        *dstPixel = (v)?(v/totalW):0;
```

Comment:
Could you insert a space before/after ? and :?

Reply:
https://github.com/blue-oil/blueoil/pull/611/commits/5dd6c9aee43af40bfbf32b738d2b0ffb07631947#diff-37230609db7212ca72dd52076761c190L135